File: leetcode/src/main/java/dynamicprogramming/_416/Solution2.java
package dynamicprogramming._416;

public class Solution2 {
    public boolean canPartition(int[] nums) {
        int sum = 0;
        for (int num : nums) sum += num;

        // 전체 합이 홀수면 두 부분으로 나눌 수 없음
        if (sum % 2 != 0) return false;

        int target = sum / 2;
        boolean[] dp = new boolean[target + 1];
        dp[0] = true; // 0을 만드는 건 가능 (부분집합 없이)

        for (int num : nums) {
            for (int i = target; i >= num; i--) {
                dp[i] = dp[i] || dp[i - num];
            }
        }

        return dp[target];
    }
    // 30ms
}

File: leetcode/src/main/java/bitmanipulation/_76_single_number_136/Solution2.kt
package bitmanipulation._76_single_number_136

class Solution2 {
    fun singleNumber(nums: IntArray): Int {
        var result = 0
        for (num in nums) {
            // 각 수를 비트 연산자 XOR로 연산한 결과를 누적한다.
            result = result xor num
        }
        return result
    }
}